src/lib.rs

```diff
@@ -104,58 +121,51 @@ impl<D> EventSource for WaylandSource<D> {
     type Metadata = EventQueue<D>;
     type Ret = Result<usize, DispatchError>;
 
+    const NEEDS_EXTRA_LIFECYCLE_EVENTS: bool = true;
+
     fn process_events<F>(
         &mut self,
-        readiness: Readiness,
-        token: Token,
+        _: Readiness,
+        _: Token,
         mut callback: F,
     ) -> Result<PostAction, Self::Error>
     where
         F: FnMut(Self::Event, &mut Self::Metadata) -> Self::Ret,
     {
         let queue = &mut self.queue;
-        let read_guard = &mut self.read_guard;
+        if let Some(err) = self.stored_error.take() {
+            return Err(err)?;
+        }
 
-        let action = self.fd.process_events(readiness, token, |_, _| {
-            // 1. read events from the socket if any are available
-            if let Some(guard) = read_guard.take() {
-                // might be None if some other thread read events before us, concurrently
-                if let Err(WaylandError::Io(err)) = guard.read() {
-                    if err.kind() != io::ErrorKind::WouldBlock {
-                        return Err(err);
-                    }
-                }
-            }
+        // In theory, we would need to call the process_events handler on fd. However,
+        // we know that Generic's process events's call is a no-op, so we can just handle the event ourselves
```

Comment:
I find it hard to understand what you want to say with this comment, since it's not attached to anything.